### ISSUE TITLE
Feature Request: Microsoft Teams integration

### ISSUE THREAD

**Issue by robbycuenot** [aws-samples/iam-identity-center-team#139]
**Describe the bug**
Feature Request: Integrate with Microsoft Teams, similar to the Slack integration.

**Expected behavior**
Notifications for requests and approvals are sent to a slack channel, and potentially to the requester / approvers directly

**Comment by robbycuenot** [comment]
I've been working on a containerized lambda with a MS Teams Webhook to handle this. Very basic working example:

```python
import json
import os
import requests
import logging
from datetime import datetime
from dateutil import parser

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

def lambda_handler(event, context):
    # Log the received SNS message
    sns_message = event['Records'][0]['Sns']['Message']
    logger.info("Received SNS message: %s", sns_message)
    
    # Assuming the SNS message is a stringified JSON
    message = json.loads(sns_message)

    # Common details extracted from the message
    email = message.get('email', 'N/A')
    username = message.get('username', 'N/A')
    time_hours = message.get('time', 'N/A')  # Using the "time" field for duration
    accountName = message.get('accountName', 'N/A')
    role = message.get('role', 'N/A')
    justification = message.get('justification', 'N/A')

    status = message.get('status', 'N/A').lower()  # Normalize the status to lower case for comparison
    startTime = message.get('startTime', 'N/A')
    if startTime != 'N/A':
        start_time_obj = parser.parse(startTime)
        startTime = start_time_obj.strftime('%Y-%m-%dT%H:%M:%S%z')

    # Construct the message for Microsoft Teams based on status
    if status == "pending":
        summary = "AWS Temporary Elevated Access Request"
        activityTitle = "🚀 New Access Request"
    elif status == "scheduled":
        summary = "AWS Access Scheduled"
        activityTitle = "⌚ Access Scheduled"
    elif status == "expired":
        summary = "AWS Access Request Expired"
        activityTitle = "⏳ Access Request Expired"
    elif status == "ended":
        summary = "AWS Access Ended"
        activityTitle = "🔒 Access Ended"
    elif status == "granted":
        summary = "AWS Access Granted"
        activityTitle = "🔓 Access Granted"
    elif status == "approved":
        summary = "AWS Access Request Approved"
        activityTitle = "✅ Access Request Approved"
    elif status == "rejected":
        summary = "AWS Access Request Rejected"
        activityTitle = "❌ Access Request Rejected"
    elif status == "cancelled":
        summary = "AWS Access Request Cancelled"
        activityTitle = "✖ Access Request Cancelled"
    else:
        summary = "AWS Access Request Unhandled Exception"
        activityTitle = "ℹ️ Access Request Unhandled Exception"

    # Construct the Teams message
    teams_message = {
        "@type": "MessageCard",
        "@context": "http://schema.org/extensions",
        "themeColor": "0076D7",
        "summary": summary,
        "sections": [{
            "activityTitle": activityTitle,
            "activitySubtitle": f"Requested by {email} ({username})",
            "facts": [
                {"name": "Duration:", "value": f"{time_hours} hours"},
                {"name": "Account Name:", "value": accountName},
                {"name": "Role:", "value": role},
                {"name": "Start Time:", "value": startTime},
                {"name": "Justification:", "value": justification}
            ],
            "markdown": True
        }]
    }

    # Retrieve Microsoft Teams Webhook URL from environment variable
    webhook_url = os.getenv('TEAMS_WEBHOOK_URL')
    
    if not webhook_url:
        logger.error('TEAMS_WEBHOOK_URL environment variable is not set.')
        return {
            'statusCode': 400,
            'body': json.dumps('TEAMS_WEBHOOK_URL environment variable is not set.')
        }
    
    # Send the message to Microsoft Teams
    response = requests.post(webhook_url, json=teams_message, headers={"Content-Type": "application/json"})
    
    if response.status_code == 200:
        logger.info('Message sent to Microsoft Teams successfully.')
        return {
            'statusCode': 200,
            'body': json.dumps('Message sent to Microsoft Teams successfully.')
        }
    else:
        logger.error(f"Failed to send message to Microsoft Teams. Response: {response.text}")
        return {
            'statusCode': response.status_code,
            'body': json.dumps(f"Failed to send message to Microsoft Teams. Response: {response.text}")
        }
```

**Comment by tawoyinfa** [comment]
Thanks for sharing this @robbycuenot We need to figure out how to provide visibility to this code so others can reuse it or integrate it as a feature. 

Ideally I'd like that the Microsoft Teams integration be built similar to the slack notification where message can be sent as a direct message to members of the approval group or requester. 

If we figure this out I am happy to prioritise it and merge into the main code base.

**Comment by robbycuenot** [comment]
@tawoyinfa I agree, having a similar integration to slack would be ideal. This setup fills the gap for now, but is certainly not optimal. Let me know if you're able to add it to the roadmap. Thanks!

**Comment by reidca** [comment]
I would be very keen to see this integration which would help a lot of companies make this more useful.

**Comment by robbycuenot** [comment]
@reidca I've been building out a pretty comprehensive add-on for this. Will see if I can publish it at some point. It's definitely not something that would be added to this root repo, as the deployment is done with terraform and it runs on a container lambda with webhooks to MS teams but it might be a decent side-car project until the official support is released

**Comment by robbycuenot** [comment]
@tawoyinfa @reidca @duanehughes 

Here is a sample of the working MS Teams integration:

https://github.com/robbycuenot/iam-identity-center-team

**Comment by reidca** [comment]
yes, thank you for this. I will also take a look. I assume we can just add the lambda functions and set them to be a recipient of the SNS notifications then create the web hook from Teams?

**Comment by tawoyinfa** [comment]
Fantastic job @robbycuenot we will work on adapting it into the main code base at some point as well 🎉

**Comment by robbycuenot** [comment]
@reidca , yes, exactly. I've deployed it with Terraform cloud, and use GitHub actions to push the image that the lambda runs (mostly because I hate dealing with Lambda layers). A few things to note:

1. This is designed to use a single channel for all notifications
    1. Click the three dot menu on the right side of the channel
    1. Click `Connectors`
    1. On `Incoming Webhook`, click `Configure`
    
    ![image](https://github.com/aws-samples/iam-identity-center-team/assets/51327557/c9a794a9-39e6-4483-8aa6-bc76b043fe6b)
    1. Name it as you wish (I used `AWS TEAM`)
    1. Upload a logo if you wish
    
    ![white-logo](https://github.com/aws-samples/iam-identity-center-team/assets/51327557/c5fcbf15-641a-4332-803f-60520cc3c955)
    1. Click `Create`
    1. A Webhook URL will be provided below. Keep this URL secret, as anyone with it can send messages to the channel
    1. Add the Webhook URL to the Lambda, either manually or through a Terraform variable if you're using the provided TF code
    1. When testing, I recommend creating a second channel. The test suite provided will send one of each notification type and @ mention you, pulling your current user info from AWS IAM Identity Center.

1. Requesters will be @ mentioned on all status updates. Approvers will only be mentioned on the initial request. You can see this in the above screenshots, where mentions are shown in red.
1. This suite may need to be adjusted based on your organization's SSO setup, especially if it's not linked to Azure AD. The SNS notifications contain references to both SSO User Usernames, and Emails, which were different in my case (internal corporate addresses vs public facing corporate addresses). It should work in either case, but you'll see references in the code to User UPN vs User Email, with the former being the Username field and the latter being the AdditionalAttributes->Email field on the SSO user.

**Comment by robbycuenot** [comment]
> Fantastic job @robbycuenot we will work on adapting it into the main code base at some point as well 🎉

@tawoyinfa thank you! I'm looking forward to that :)

**Comment by jfevre** [comment]
@tawoyinfa is this still on the road map to be added into the main code base ?

**Comment by robbycuenot** [comment]
> @tawoyinfa is this still on the road map to be added into the main code base ?

I hope it is at some point. I will note, however, that MS deprecating Webhooks has made this more of a pain in the butt than it used to be. You have to create a power automate flow, which has to be associated with a real user in order to post messages. I've gotten it working in my org, but it's quite a few extra steps compared to what it used to be. We've been using this for close to a year now

**Comment by robbycuenot** [comment]
@tawoyinfa I recently discovered this AWS service which could make the MS Teams integration much easier: https://docs.aws.amazon.com/chatbot/latest/adminguide/teams-setup.html

Could you re-open this issue?

**Comment by tawoyinfa** [comment]
> I hope it is at some point. I will note, however, that MS deprecating Webhooks has made this more of a pain in the butt than it used to be. You have to create a power automate flow, which has to be associated with a real user in order to post messages. I've gotten it working in my org, but it's quite a few extra steps compared to what it used to be. We've been using this for close to a year now

@robbycuenot with the power automate implementation you deployed in house, does the message show up as coming from the TEAM app/bot or does it show up as coming from the associated user ?

**Comment by robbycuenot** [comment]
> 
> > I hope it is at some point. I will note, however, that MS deprecating Webhooks has made this more of a pain in the butt than it used to be. You have to create a power automate flow, which has to be associated with a real user in order to post messages. I've gotten it working in my org, but it's quite a few extra steps compared to what it used to be. We've been using this for close to a year now
> 
> @robbycuenot with the power automate implementation you deployed in house, does the message show up as coming from the TEAM app/bot or does it show up as coming from the associated user ?
> 
> 
> 

@tawoyinfa it shows as coming from the associated user, which is very unfortunate. The only way to work around this is to create a service account with a teams license and associate the flow with that. I am hoping the AWS Chatbot app allows us to build a similar flow without the service account, but I haven't tested it yet

**Comment by mdgm88** [comment]
Any update on this?

**Comment by mdgm88** [comment]
We don't want to deploy customisations and then have to assess what breaks with updates to newer versions of TEAM. I guess I'll wait and see if this gets implemented in the future.
